---
## pregunta uno
Se dispone de tres listas: nombres de clientes, edades y montos de compra en CLP.
### Tarea:

1.  **Construya un DataFrame en pandas asegurando:**
    *   Tipos de datos correctos (dtype)
    *   Indexación personalizada (ID de cliente)

2.  **Explique:**
    *   La diferencia entre DataFrame y Series
    *   Cómo influye el tipo de dato en la eficiencia computacional

  ---

   ## solucion
  construccion de dt en pandas
#datos:




In [14]:
import pandas as pd

# este DataFrame representa clientes con su edad y monto de compras en CLP.
# se asigna un índice personalizado para identificar cada cliente.

# ejemplo de datos
nombre_clientes = ['Juan', 'Maria', 'Pedro', 'Ana', 'Luis']
edades = [30, 25, 28, 79, 28]
monto_compras_CLP = [100000, 200000, 150000, 300000, 250000]

#crear diccionarios
datos= {
    'nombre': nombre_clientes,
    'edad': edades,
    'monto_compra_CLP': monto_compras_CLP
}

#crear dataframe
df = pd.DataFrame(datos)

#asignar un ID a clientes de forma personalizada
df.index = [f"clientes_{i+1:03}" for i in range(len(df))]

print("\nse ha creado el siguiente dataframe:\n")
print(df)

print("\nTipos de datos del DataFrame:\n")
print(df.dtypes)

print("\nResumen estadístico:\n")
print(df.describe())



se ha creado el siguiente dataframe:

             nombre  edad  monto_compra_CLP
clientes_001   Juan    30            100000
clientes_002  Maria    25            200000
clientes_003  Pedro    28            150000
clientes_004    Ana    79            300000
clientes_005   Luis    28            250000

Tipos de datos del DataFrame:

nombre              object
edad                 int64
monto_compra_CLP     int64
dtype: object

Resumen estadístico:

            edad  monto_compra_CLP
count   5.000000          5.000000
mean   38.000000     200000.000000
std    22.989128      79056.941504
min    25.000000     100000.000000
25%    28.000000     150000.000000
50%    28.000000     200000.000000
75%    30.000000     250000.000000
max    79.000000     300000.000000


## explicacion dtype e indexacion

* **dtype:** la librería pandas (pd) infiere automáticamente los tipos de datos. Esto se hace para optimizar la memoria y el rendimiento. Por ejemplo, `int64` corresponde a enteros y `float` a números decimales.

* **indexacion (ID):** se agregó un formato `cliente_001` para identificar de manera única cada fila de clientes. Esto ayuda a ordenar, buscar y acceder a los datos de forma más clara.

#### diferencia entre DataFrame y Series

como vimos en clases, se explicó el significado de PanDaS:

* Pan: hace referencia a panel  
* Da: nos habla de data o datos  
* S: hace referencia a series  

**Series:** una serie es un array de una dimensión y es capaz de contener cualquier tipo de dato. En Excel sería como una columna que almacena datos. Cada elemento de una serie tiene un "índice", conocido como etiqueta.

**DataFrame:** es una estructura de datos de dos dimensiones etiquetada con columnas. Podemos entenderlo como una tabla SQL o una hoja de Excel. Es una colección de series que comparten el mismo índice.  **(Un ejemplo sería el que construimos anteriormente)**

## Cómo influye el tipo de dato en la eficiencia computacional

El tipo de dato `dtype` tiene varios impactos en la eficiencia computacional:

**Uso de memoria:**

* **Tipos de datos más pequeños:** existen diferentes tipos como `int8`, `int32`, `int64`. Mientras más pequeño el tipo, menor uso de memoria RAM, lo que puede hacer que las operaciones sean más rápidas.

* **Category dtype:** para columnas con un número limitado de valores únicos, convertir el tipo de dato a `category` puede reducir bastante el uso de memoria y acelerar operaciones como `groupby`.

**Velocidad de operaciones:**

* **Operaciones numéricas:** las operaciones matemáticas en columnas numéricas (`int`, `float`) son mucho más rápidas que en columnas de tipo `object` (texto). Esto se debe a que pandas y NumPy están optimizados para trabajar con datos numéricos.

* **Comparaciones y filtrado:** son más eficientes con datos numéricos o booleanos que con texto, especialmente si las cadenas son largas.

* **Missing values:** el manejo de valores nulos también influye. Por ejemplo, en datos numéricos se usa `NaN`, mientras que en `object` puede usarse `None` o texto vacío, lo que puede afectar el rendimiento.

**Compatibilidad y errores:**

* **Operaciones inválidas:** realizar operaciones numéricas en columnas con tipos incorrectos puede generar errores y obliga a convertir datos, lo que consume tiempo.

* **Consistencia:** mantener tipos de datos consistentes ayuda a evitar errores y asegura que las funciones de `pandas` funcionen correctamente.

---
##pregunta 2

Dado un DataFrame de ventas con columnas: Cliente, Producto, Monto, Fecha: (cree uno con al menos 20 registros)

### Tarea:

*   Seleccione todos los registros donde:
    *   Monto > 50000 y Producto == "Laptop"
*   Modifique el valor del Monto en un 10% de aumento para dichos registros
*   Explique por qué `loc` es preferible frente a indexación tradicional en este caso
---

##solucion
### creacion del Df de ventas

importar librerias `pandas` y `numpy`


In [13]:
import pandas as pd
import numpy as np

#se crean datos aleatorios
np.random.seed(42)

#usamos 25 registros para continuar
n_registros = 25
#usamos un for para ir por cada cliente
clientes = [f'Cliente_{i}' for i in range(1, n_registros + 1)]
#asignamos los productos mediante una lista
productos_lista = ['Laptop', 'Teclado', 'Mouse', 'Monitor', 'Webcam']

# Crea un rango de fechas con las siguientes características:
# start: Comienza el 1 de enero de 2023
# periods: Genera exactamente la cantidad de fechas que indica la variable 'n_registros'
# freq='D': El salto entre cada fecha es de un día (Diario)
fechas = pd.to_datetime(pd.date_range(start='2023-01-01', periods=n_registros, freq='D'))

#aignamos datos a cada fila con su respectiva etiqueta
df_ventas = pd.DataFrame({
    'Cliente': np.random.choice(clientes, size=n_registros),
    'Producto': np.random.choice(productos_lista, size=n_registros),
    'Monto': np.random.randint(10000, 100000, size=n_registros),
    'Fecha': fechas
})
# hacemos que los apartados de laptop esten fijo (para tener dos >50.000)
df_ventas.loc[5, ['Producto', 'Monto']] = ['Laptop', 60000]
df_ventas.loc[10, ['Producto', 'Monto']] = ['Laptop', 75000]

#imprime el df por pantalla
print("DataFrame de Ventas (Primeros 12 registros):\n")
print(df_ventas.head(12))


DataFrame de Ventas (Primeros 12 registros):

       Cliente Producto  Monto      Fecha
0    Cliente_7   Laptop  83969 2023-01-01
1   Cliente_20  Monitor  53001 2023-01-02
2   Cliente_15  Teclado  86552 2023-01-03
3   Cliente_11   Webcam  33897 2023-01-04
4    Cliente_8  Monitor  78148 2023-01-05
5   Cliente_21   Laptop  60000 2023-01-06
6    Cliente_7   Laptop  58555 2023-01-07
7   Cliente_19    Mouse  27159 2023-01-08
8   Cliente_23    Mouse  90077 2023-01-09
9   Cliente_11  Teclado  45920 2023-01-10
10  Cliente_11   Laptop  75000 2023-01-11
11  Cliente_24  Monitor  79479 2023-01-12


### Se seleccionan los registros donde el `Monto` sea mayor a 50000 y el `Producto` sea "Laptop". Luego, aumentaremos el `Monto` de estos registros en un 10%.

In [22]:
# Definir la condición
condicion = (df_ventas['Monto'] > 50000) & (df_ventas['Producto'] == 'Laptop')

# Guardamos una copia ANTES de modificar (importante para comparar)
df_filtrado_antes = df_ventas[condicion].copy()

print("\nRegistros que cumplen la condición ANTES del aumento:\n")
print(df_filtrado_antes)

# Aplicar el aumento del 10% usando .loc
# .loc[filas, columna] garantiza que modificamos el DataFrame original
df_ventas.loc[condicion, 'Monto'] *= 1.10

print("\nRegistros que cumplen la condición DESPUÉS del aumento (10% extra):\n")
print(df_ventas.loc[condicion])


Registros que cumplen la condición ANTES del aumento:

       Cliente Producto          Monto      Fecha
0    Cliente_7   Laptop  163631.826170 2023-01-01
5   Cliente_21   Laptop  116923.026000 2023-01-06
6    Cliente_7   Laptop  114107.129791 2023-01-07
10  Cliente_11   Laptop  146153.782500 2023-01-11
15  Cliente_24   Laptop  169906.695232 2023-01-16
20  Cliente_24   Laptop  108656.568062 2023-01-21
23   Cliente_2   Laptop  147512.038319 2023-01-24

Registros que cumplen la condición DESPUÉS del aumento (10% extra):

       Cliente Producto          Monto      Fecha
0    Cliente_7   Laptop  179995.008787 2023-01-01
5   Cliente_21   Laptop  128615.328600 2023-01-06
6    Cliente_7   Laptop  125517.842770 2023-01-07
10  Cliente_11   Laptop  160769.160750 2023-01-11
15  Cliente_24   Laptop  186897.364755 2023-01-16
20  Cliente_24   Laptop  119522.224868 2023-01-21
23   Cliente_2   Laptop  162263.242151 2023-01-24


####explicación: ¿Por qué loc es preferible frente a la indexación tradicional?

El uso de `.loc` es la forma estándar y recomendada en pandas por las siguientes razones:

* **Evita el error de copia:** la indexación tradicional como `df[condicion]['Monto']` realiza lo que se conoce como "indexación encadenada". En este caso, pandas no puede asegurar si se está trabajando sobre una copia o una vista del DataFrame original, lo que puede provocar que los cambios no se guarden correctamente.

* **Claridad:** con `.loc[filas, columnas]` se indica explícitamente qué filas y qué columna se desean modificar dentro del DataFrame original.

* **Consistencia:** `.loc` es más predecible, ya que siempre modifica directamente el DataFrame, evitando comportamientos inesperados.

---
##pregunta 3
Considere un dataset con al menos 100 observaciones:

**Extraiga:**
* Las primeras 10 filas
* Las últimas 5 columnas
* Un subconjunto rectangular (filas 20 a 40, columnas 2 a 5)
* Analice:
  Diferencias conceptuales entre loc e iloc
  
  Riesgos de errores silenciosos en indexación posicional


**dataset obtenido de kaggle**

Para este análisis, se ha utilizado el dataset **World Happiness Report**, el cual contiene información sobre los niveles de felicidad en más de 150 países.

---


In [28]:
import pandas as pd
#importamos los datos para seguir con el analisis
url = "https://raw.githubusercontent.com/PhilippeCodes/World-Happiness-Report-Data-Analysis/master/World%20Happiness%20Report.csv"
df = pd.read_csv(url )

#pedimos ver las primeras 10 filas para ver de que tratan los datos
primeras_10 = df.head(10)
#pedimos ver las ultimas filas para ver de que tratan los datos
ultimas_5_col = df.iloc[:, -5:]
subconjunto = df.iloc[20:41, 2:6]

#mostrar todo lo obtenido
print("Primeras 10 filas:\n")
print(primeras_10)

print("\nÚltimas 5 columnas:\n")
print(ultimas_5_col.head())

print("\nSubconjunto rectangular:\n")
print(subconjunto)
print("\nDimensiones del dataset:", df.shape)

Primeras 10 filas:

       Country  Happiness Rank  Happiness Score   Economy    Family    Health   Freedom  Generosity  Corruption  Dystopia  Job Satisfaction          Region
0       Norway               1            7.537  1.616463  1.533524  0.796667  0.635423    0.362012    0.315964  2.277027              94.6  Western Europe
1      Denmark               2            7.522  1.482383  1.551122  0.792566  0.626007    0.355280    0.400770  2.313707              93.5  Western Europe
2      Iceland               3            7.504  1.480633  1.610574  0.833552  0.627163    0.475540    0.153527  2.322715              94.5  Western Europe
3  Switzerland               4            7.494  1.564980  1.516912  0.858131  0.620071    0.290549    0.367007  2.276716              93.7  Western Europe
4      Finland               5            7.469  1.443572  1.540247  0.809158  0.617951    0.245483    0.382612  2.430182              91.2  Western Europe
5  Netherlands               6            7.

## diferencia entre loc e iloc
Como revisamos antes en DataFrames, existen dos formas para extraer datos:

*  **loc (Localización por etiqueta):** Busca la información utilizando los nombres explícitos (etiquetas) asignados a las filas o columnas. Una característica clave es que al usar rangos (ej: `loc['A':'C']`), el último elemento del rango **sí** se incluye en el resultado final.

*  **iloc (Localización posicional):** Busca la información basándose estrictamente en su posición numérica entera (0, 1, 2...), tal como funcionan las listas tradicionales. A diferencia de `loc`, al usar rangos (ej: `iloc[0:3]`), el límite superior **no** se incluye (solo traería las posiciones 0, 1 y 2).

## Riesgos de errores silenciosos en indexación posicional

El uso de (`iloc`) tiene ciertos riesgos si no tenemos control absoluto sobre la estructura de los datos :

**¿Qué es un error silencioso?**
*  **Concepto:** Ocurre cuando un bloque de código se ejecuta "exitosamente" sin lanzar ninguna alerta, detención o mensaje de error en pantalla, pero el resultado de los datos es incorrecto.

**El problema al depender de iloc:**
*  **Sensibilidad a cambios de estructura:** `iloc` asume que el orden del dataset es totalmente estático. Si el dataset cambia su estructura (por ejemplo, si se agrega una columna nueva al principio o se reordenan), el código seguirá funcionando pero **seleccionará datos incorrectos**. Esto se llama error silencioso porque no arroja un mensaje de error, pero los resultados del análisis serán falsos.


---
##pregunta 4: análisis de Comportamiento de Clientes

**Realice:**

* Filtrado por múltiples condiciones (edad, región, nivel de ingreso)
* Segmentación por percentiles (ej: top 25% de ingresos)
* Explique cómo el slicing permite construir variables derivadas útiles para modelos predictivos


Se analiza un dataset de comportamiento de clientes para identificar segmentos específicos mediante filtrado y uso de percentiles.

---


In [30]:
import pandas as pd
import numpy as np

#usamos la url que es de donde obtenemos los datos para seguir con el analisis
url = "https://raw.githubusercontent.com/tirthajyoti/Machine-Learning-with-Python/master/Datasets/Mall_Customers.csv"
df = pd.read_csv(url )
regiones = ['Norte', 'Sur', 'Este', 'Oeste']
#datos aleatorios
np.random.seed(42)
# punto importante es añadir una columna region, ya que no existe
df['Region'] = np.random.choice(regiones, size=len(df))

#filtrado multiple bajo ciertas condiciones
condicion = (df['Age'] > 30) & (df['Region'] == 'Norte') & (df['Annual Income (k$)'] > 50)
#se aplica la condicion
clientes_filtrados = df.loc[condicion]

#segmentacion por percentiles
umbral = df['Annual Income (k$)'].quantile(0.75)
#se filtra el data set original para que cumplan la condicion
top_clientes = df.loc[df['Annual Income (k$)'] >= umbral]

#imprimir los primeros 5 resultados
print("Clientes Filtrados (Primeros 5):\n")
print(clientes_filtrados.head())
print(f"\nClientes Top 25% Ingresos (Umbral: {umbral}k$):\n")
print(top_clientes.head())
#conteo clientes
print("\nCantidad de clientes filtrados:", len(clientes_filtrados))
print("Cantidad de clientes top 25%:", len(top_clientes))



Clientes Filtrados (Primeros 5):

     CustomerID  Gender  Age  Annual Income (k$)  Spending Score (1-100) Region
74           75    Male   59                  54                      47  Norte
82           83    Male   67                  54                      41  Norte
116         117  Female   63                  65                      43  Norte
119         120  Female   50                  67                      57  Norte
121         122  Female   38                  67                      40  Norte

Clientes Top 25% Ingresos (Umbral: 78.0k$):

     CustomerID  Gender  Age  Annual Income (k$)  Spending Score (1-100) Region
148         149  Female   34                  78                      22  Norte
149         150    Male   34                  78                      90  Oeste
150         151    Male   43                  78                      17   Este
151         152    Male   39                  78                      88   Este
152         153  Female   44            

##Explicación: Slicing y Variables Derivadas

El **slicing** (o recorte de datos) es fundamental para construir variables derivadas útiles en modelos predictivos por las siguientes razones:

1.  **Creación de Etiquetas de Segmento**: Permite dividir una variable continua (como el ingreso) en categorías (ej: "Ingreso Alto", "Ingreso Medio", "Ingreso Bajo"). Estas etiquetas ayudan a los modelos a capturar patrones no lineales.
2.  **Aislamiento de Comportamientos**: Al recortar datos por periodos de tiempo o regiones, podemos crear variables como "Gasto promedio en los últimos 3 meses", lo cual es un predictor mucho más potente que el gasto total histórico.
3.  **Reducción de Ruido**: Permite eliminar valores atípicos (outliers) mediante el recorte de percentiles extremos, mejorando la precisión y estabilidad de los modelos de Machine Learning.

---
##pregunta 5: medidas de tendencia central

Calcule:
* Media, mediana y moda
* Compare:
Resultados usando pandas vs cálculo manual

* Interprete:
  Qué ocurre en presencia de outliers

  Cuándo es preferible cada medida

se analiza el dataset Titanic para comparar los métodos de cálculo y entender el impacto de los valores extremos.

---



In [32]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url )
data = df['Fare'].dropna() # seleccionar columna de tarifa para trabjar en ella

print(data.head(5)) # lo suamos para visualizar los datos

# usando la libreria pandas(pd) es relativamente sencillo calcular las 3
media_pd = data.mean()
mediana_pd = data.median()
# usamos [0] porque mode() puede devolver más de un valor
moda_pd = data.mode()[0]

# calculo manual usando fuunciones y logica:
# media
media_manual = sum(data) / len(data)

# mediana
data_sorted = sorted(data)
n = len(data_sorted)
if n % 2 == 0:
    mediana_manual = (data_sorted[n//2 - 1] + data_sorted[n//2]) / 2
else:
    mediana_manual = data_sorted[n//2]

# calculo de la moda
frecuencias = {}
for valor in data:
    frecuencias[valor] = frecuencias.get(valor, 0) + 1

# buscamos la frecuencia con el valor maximo
moda_manual = max(frecuencias, key=frecuencias.get)

print("\ncomparación de resultados")
print(f"Media   = Pandas: {media_pd:.2f} | Manual: {media_manual:.2f}")
print(f"Mediana = Pandas: {mediana_pd:.2f} | Manual: {mediana_manual:.2f}")
print(f"Moda    = Pandas: {moda_pd:.2f}  | Manual: {moda_manual:.2f}")

print("\nResumen estadístico:\n")
print(data.describe())


0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
Name: Fare, dtype: float64

comparación de resultados
Media   = Pandas: 32.20 | Manual: 32.20
Mediana = Pandas: 14.45 | Manual: 14.45
Moda    = Pandas: 8.05  | Manual: 8.05

Resumen estadístico:

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64


#### Comparación uso de: pandas vs manual

*  **Resultados:** Ambos métodos arrojan exactamente los mismos valores, lo que valida la lógica matemática detrás de las funciones de Pandas.
*  **Eficiencia:** El cálculo manual requiere iteraciones explícitas (bucles `for`, ordenamiento manual), lo cual es ineficiente en grandes volúmenes de datos. Pandas utiliza operaciones vectorizadas, haciendo que el cálculo sea casi instantáneo con menos líneas de código.

## Interpretación de Medidas de Tendencia Central



### ¿Qué ocurre en presencia de outliers?
En este dataset, el valor máximo es **512.33**, mientras que la mediana es solo **14.45**.

* La **media** se ve "arrastrada" hacia arriba por los valores altos (outliers), dando una impresión de que el pasaje promedio era más caro de lo que realmente era.
* La **mediana** se mantiene estable, mostrando que el 50% de los pasajeros pagó menos de 14.45, lo cual representa mejor la realidad del conjunto de datos.

### ¿Cuándo es preferible cada medida?
1.  **Media**: Es preferible cuando los datos son **simétricos** y no tienen valores extremos (ej: alturas de personas).
2.  **Mediana**: Es la mejor opción cuando hay **outliers** o datos sesgados (ej: ingresos, precios de viviendas), ya que no se ve afectada por valores inusualmente altos o bajos.
3.  **Moda**: Es útil para variables **categóricas** o cuando queremos saber cuál es el valor más común o frecuente en una serie (ej: la talla de zapato más vendida).


---
## pregunta 6: Tabla y frecuencias

Construya:
* Tabla de frecuencia absoluta
* Frecuencia relativa
* Frecuencia acumulada
* Analice:
 Cómo estas tablas pueden apoyar decisiones de negocio

la construcción y análisis de tablas de frecuencia de una variable

---

In [9]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# se usa clase de billete para analisis
frec_abs = df['Pclass'].value_counts().sort_index()
frec_rel = df['Pclass'].value_counts(normalize=True).sort_index()
frec_acum = frec_abs.cumsum()

# se construye la tabla
tabla_frecuencias = pd.DataFrame({
    'Frecuencia Absoluta': frec_abs,
    'Frecuencia Relativa (%)': (frec_rel * 100).round(2),
    'Frecuencia Acumulada': frec_acum
})

tabla_frecuencias.index.name = 'Clase de Pasajero'

# esto es algo que vimos en una clase cuando se mostro el histograma o en frecuencias
# no alcanzaba el display y se edito de una manera que se pudiese visualizar completo
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

print("tabla de frecuencias\n")
print(tabla_frecuencias)

tabla de frecuencias

                   Frecuencia Absoluta  Frecuencia Relativa (%)  Frecuencia Acumulada
Clase de Pasajero                                                                    
1                                  216                    24.24                   216
2                                  184                    20.65                   400
3                                  491                    55.11                   891


## analisis: uso de tablas de frecuencia en decisiones de negocio

las tablas de frecuencia permiten entender rapidamente como se distribuyen los datos dentro de una variable, en este caso la clase de pasajero (`Pclass`).

* **frecuencia absoluta:** muestra cuántos pasajeros hay en cada clase. esto permite identificar qué segmento es más común dentro del dataset.

* **frecuencia relativa:** indica el porcentaje que representa cada clase respecto al total. esto es útil para comparar proporciones y entender el peso real de cada grupo.

* **frecuencia acumulada:** permite ver cómo se van sumando los datos hasta cierto punto. aunque en variables categóricas no siempre es tan relevante, puede ayudar a ver concentraciones.

### aplicacion en decisiones de negocio

* **segmentacion de clientes:** permite identificar cuál es el grupo predominante (por ejemplo, más pasajeros en clase baja), lo que ayuda a enfocar estrategias en ese segmento.

* **definicion de precios:** si la mayoría pertenece a clases más económicas, se pueden ajustar precios o promociones para ese grupo.

* **optimizacion de recursos:** entender la distribución ayuda a asignar mejor recursos (espacio, servicios, atención) según la demanda real.

* **deteccion de oportunidades:** si un segmento tiene baja representación, puede ser una oportunidad de crecimiento o expansión.

en resumen, estas tablas permiten transformar datos en informacion util para tomar decisiones más informadas y reducir la incertidumbre.

---
###pregunta 7
### ya que es el apartado mas importante intentare comentar cada linea

Construya un flujo completo:
* Limpieza de datos
* Filtrado con loc e iloc
* Cálculo de estadísticos descriptivos
* Construcción de tablas de frecuencia
* Generación de insights
* Redacte un breve informe técnico interpretando los resultados

---

In [12]:
import pandas as pd
import numpy as np

# usamos un dataset Titanic adaptado a un contexto de Marketing
# usamos un url para tener el df
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
# usamos df.read para poder leer el archivo.csv
df = pd.read_csv(url )

#pritn para visualizar la carga
print("Dataset cargado exitosamente\n")
print(f"Total de registros: {len(df)}")

# Eliminamos filas con valores nulos en columnas
df = df.dropna(subset=['Age', 'Fare', 'Embarked'])

# Cambiamos el nombre a minusculas para facilitar la busqueda
df.columns = [col.strip().lower() for col in df.columns]

# filtrado con .iloc y .loc
# Uso de loc: Clientes mayores de 30 años que pagaron una tarifa alta (Segmento premium)
filtro_loc = df.loc[(df['age'] > 30) & (df['fare'] > 50)]

# Uso de iloc: Seleccionar las primeras 5 filas y las primeras 4 columnas para inspección
filtro_iloc = df.iloc[0:5, 0:4]

#print para visualizar el filtrado
print("\nResultado filtrado loc (Segmento premium)\n")
print(filtro_loc[['age', 'fare', 'sex']].head())

# caluclos estadisticos
# Obtenemos promedios, valores mínimos y máximos de edad y gasto (Fare)
estadisticos = df[['age', 'fare']].describe()

#print para ver los resultados del analisis
print("\nEstadísticos descriptivos (Edad y gasto)\n")
print(estadisticos)

# construir tablas de frecuencia
# Tabla de frecuencia absoluta para la variable 'pclass' (Clase/nivel socioeconómico)
frec_absoluta = df['pclass'].value_counts().sort_index()

# Tabla de frecuencia relativa (%)
frec_relativa = (df['pclass'].value_counts(normalize=True) * 100).sort_index().round(2)

#tabla asociada a sus respectivos nombres
tabla_frecuencia = pd.DataFrame({
    'Frecuencia absoluta': frec_absoluta,
    'Frecuencia relativa (%)': frec_relativa
})
#cambio de index por nivel clase
tabla_frecuencia.index.name = 'Nivel/Clase'

print("\nTabla de Frecuencia: Niveles Socioeconómicos\n")
print(tabla_frecuencia)

# GENERACIÓN DE INSIGHTS
# Insight 1: Gasto promedio por género
gasto_genero = df.groupby('sex')['fare'].mean()

# Insight 2: Tasa de conversión (survived) general
tasa_conversion = df['survived'].mean() * 100

print("\nINSIGHTS ESTRATÉGICOS\n")
print(f"1. El gasto promedio (Fare) de las mujeres es {gasto_genero['female']:.2f} vs {gasto_genero['male']:.2f} de los hombres.")
print(f"2. La tasa de conversión general de la campaña es del {tasa_conversion:.2f}%.")

Dataset cargado exitosamente

Total de registros: 891

Resultado filtrado loc (Segmento premium)

     age     fare     sex
1   38.0  71.2833  female
3   35.0  53.1000  female
6   54.0  51.8625    male
35  42.0  52.0000    male
52  49.0  76.7292  female

Estadísticos descriptivos (Edad y gasto)

              age        fare
count  712.000000  712.000000
mean    29.642093   34.567251
std     14.492933   52.938648
min      0.420000    0.000000
25%     20.000000    8.050000
50%     28.000000   15.645850
75%     38.000000   33.000000
max     80.000000  512.329200

Tabla de Frecuencia: Niveles Socioeconómicos

             Frecuencia absoluta  Frecuencia relativa (%)
Nivel/Clase                                              
1                            184                    25.84
2                            173                    24.30
3                            355                    49.86

INSIGHTS ESTRATÉGICOS

1. El gasto promedio (Fare) de las mujeres es 47.33 vs 27.27 de los homb

## informe tecnico interpretativo

### A. limpieza de datos
se eliminaron filas con valores nulos en variables importantes como edad, tarifa y embarque. esto es importante porque evita errores en los cálculos y hace que los resultados sean más confiables.  
también se cambiaron los nombres de las columnas a minúsculas para trabajar más cómodo y evitar errores al escribir.

### B. segmentación
se aplicaron dos tipos de filtrado:

* **loc:** se utilizó para identificar un segmento específico de clientes (mayores de 30 años y con tarifa alta). este grupo representa clientes con mayor capacidad de gasto, por lo que puede ser interesante para estrategias más enfocadas.

* **iloc:** se usó para ver una parte del dataset de forma rápida (primeras filas y columnas). esto sirve más para revisar que los datos estén bien estructurados.

### C. analisis de datos
* **estadisticos:** se analizaron variables como edad y tarifa. esto permite entender el comportamiento general de los clientes, como promedios, valores altos y dispersión.

* **frecuencias:** la distribución de clases muestra que la mayoría de los clientes pertenece a la clase 3. esto indica que el mayor volumen está en un segmento más económico.

### D. insights
* **diferencia de gasto:** se observa que existe una diferencia en el gasto promedio entre hombres y mujeres. esto puede servir para enfocar mejor campañas o estrategias.

* **tasa de conversión:** el porcentaje de sobrevivientes se interpreta como una tasa general. aunque no es un contexto real de negocio, sirve para entender proporciones dentro del dataset.

**conclusion**

el analisis permite ver que no todos los clientes tienen el mismo comportamiento. existen segmentos con mayor gasto y otros con mayor volumen.  
esto es clave para tomar decisiones, ya que permite enfocar mejor los recursos y entender a qué tipo de cliente se le puede sacar mayor valor.